# GEOS 505: Problem Set 1

__Instructions__: Complete the two problems below. One is a coding problem, where you will review and comment existing code. Where you are asked to provide descriptive text and answer questions, please do so via well formatted Markdown cells below the problem.  

__Due Date and Time__: September 19, 2025 at 5:00 PM MT

__Turn In Via__: Commit and push your complete notebook to your personal GitHub repository for the class, and submit the URL for notebook via Canvas. 

## Problem 1

### 1. Problem Background:

As a new graduate student, your research is examining the performance of 5-day precipitation forecasts at a number of Snotel sites in the Upper Boise River Basin. The forecasts are generated by a research center at a university in the western United States using the [Weather Research and Forecasting](https://www2.mmm.ucar.edu/wrf/users/) model. They run a 360 hr (i.e., 15 day) forecast using WRF every day at the top of every hour for the entire western United States, and puts the output you need on an Amazon Web Services (AWS) S3 bucket that is freely available.  

File pattern:
WRF_fcst_YYYYMMDD_HHz_VVV.nc

Where:
- YYYY is the year of the forecast date
- MM is the month of the forecast date
- DD is the day of the forecast
- HH is the hour the forecast was initiated (0-23)
- VVV is the valid hour of the forecast from the time of initiation in hours (0-360 hr in 3 hr increments). So 000 is the initial hour of the forecast, 003 would be the 3rd hour, 006 the 6th, and so forth.
- 'WRF_fcst_' is the prefix of the file name
- '.nc' is the file extension (this stands for NetCDF file - a real file format)

### 2. Instructions: 

1. Write a function that takes as input the forecast year, month, day, and hour, as well as a maximum valid hour and returns a list of file names that contain the forecast data
2. Test your code for three days (2025-03-01, 2025-04-01, and 2025-05-01), for forecast hour 6, and a maximum valid hour of 120 hours. Print the results of the returned list to the screen and verify manually. 

### 3. Concepts Assessed:

- String handling
- For loops
- Basic functions

In [1]:
## Your code goes here:
prefix = "WRF_fcst_"
suffix = ".nc"
#month day dictionary helps create a flexible reference for month loop
month_day = {1:31, 2:28, 3:31, 4:30, 5:31, 6:30, 7:31, 8:31, 9:30, 10:31, 11:30, 12:31} 
#for input, define a vector: year, month, day, vvv
years = [2025]
months = [3,4,5]
hours = [6]
vvvs = [120] 
max_valid_hour = 120

def DataPull(years,months,hours,vvvs):
  for year in years:
    for month in months:
      for day in range(1,month_day[month] + 1): #month_day + 1 so that the whole range is included and we dont chop off the last day!):
        for hour in hours:
          for vvv in range(0,361, 3):
            if day <= 1 and vvv <= max_valid_hour:
              print(f'{prefix}{year}{month:02d}{day:02d}_{hour:02d}z_{vvv:03d}{suffix}')
DataPull(years,months,hours,vvvs)
          
          #now I need to have accomodation for leap years!
            #something like... if year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
                                      #month_day[2]= 29


WRF_fcst_20250301_06z_000.nc
WRF_fcst_20250301_06z_003.nc
WRF_fcst_20250301_06z_006.nc
WRF_fcst_20250301_06z_009.nc
WRF_fcst_20250301_06z_012.nc
WRF_fcst_20250301_06z_015.nc
WRF_fcst_20250301_06z_018.nc
WRF_fcst_20250301_06z_021.nc
WRF_fcst_20250301_06z_024.nc
WRF_fcst_20250301_06z_027.nc
WRF_fcst_20250301_06z_030.nc
WRF_fcst_20250301_06z_033.nc
WRF_fcst_20250301_06z_036.nc
WRF_fcst_20250301_06z_039.nc
WRF_fcst_20250301_06z_042.nc
WRF_fcst_20250301_06z_045.nc
WRF_fcst_20250301_06z_048.nc
WRF_fcst_20250301_06z_051.nc
WRF_fcst_20250301_06z_054.nc
WRF_fcst_20250301_06z_057.nc
WRF_fcst_20250301_06z_060.nc
WRF_fcst_20250301_06z_063.nc
WRF_fcst_20250301_06z_066.nc
WRF_fcst_20250301_06z_069.nc
WRF_fcst_20250301_06z_072.nc
WRF_fcst_20250301_06z_075.nc
WRF_fcst_20250301_06z_078.nc
WRF_fcst_20250301_06z_081.nc
WRF_fcst_20250301_06z_084.nc
WRF_fcst_20250301_06z_087.nc
WRF_fcst_20250301_06z_090.nc
WRF_fcst_20250301_06z_093.nc
WRF_fcst_20250301_06z_096.nc
WRF_fcst_20250301_06z_099.nc
WRF_fcst_20250

I wanted this code to be more flexible, so I looped through each variable instead of printing just the prompted variables. In order to make this loop a bit more broad, I would alter the 7th line of the function so that looser boundaries were being placed on the function itself. I need to think about how to store and reference a vector with multiple values as placeholders for different variables... it feels like a dangerous route for misinterpretation if someone else were to examine the code. Is there a way for the four variables I reference in the DataPull function to be stored in a single vector and then use that vector to run the function? Is that even necessary? The way I present it here feels obvious, but simple (and works!).

## Problem 2

### 1. Problem Background

You just started grad school and inhereted some code from the final chapter of a previous student's thesis. That last chapter was completed in about 1 month and your advisor has said they want you to follow up on the precipitation analysis the student has done with a new version of the precipitation dataset that was just released. 

As you dig through their code to make it work, you realize it's not commented well at all. Additionally, the student is now working Goldman Sachs working on CliFi (climate finance) problems and, well, you can't afford their hourly rate to get some help! In order to get the data, you would need to modify the name of the files, because the revised dataset uses a different naming convention. The previous student's code to create a list of the names of files to download and analyze is below. 

### 2. Instructions

1. In your own words, what does the code do? Is it correct? 
2. Go through the old student's code and add comments describing what the code is doing. Comments should be thorough enough that they're helpful for you, but also the student that comes after you. 
3. Are there some things in the code that you would do to edit the code for clarity (i.e., to make it more readible)? Make a list of these things, but do not implement them. 

### 3. Concepts Assessed:

- Peer review
- Commenting code
- String handling
- For loops
- Functions

In [1]:
def IsLeapYear(year):
    
    if (year % 4 == 0): #divisible by 4 with no remainder?
        if (year % 500 == 0): #divisible by 500 with no remainder? shouldnt this be 400? 2400 isnt divisible by 500, but is a leap year
            return True
        elif (year % 100 == 0): #leap years are absurd that they don't happen on centuries that are not divisible by 400
            return False #EX: 1900 is not a leap year because 1900/400 is 4.75
        else:
            return True
    else:
        return False # returns for ifs and elif conditionals
    
def ReturnFileList(file_base, file_ext, start_yr, end_yr):
    
    DiM = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31] #day in month - this is like my month_day dictionary above. I am describing it as a numbered list, this is just a vector?
    year = start_yr #initialize inputs
    files = [] #create blank vector to store output
    
    for i in range(end_yr - start_yr + 1): #calculates how many times to loop... 2000-1998 +1 = 3, this is so it is inclusive of both start and end year instead of the difference between them
        for j in range(12): #range of 12 works because DiM[0] is January
            if (j+1==2) and IsLeapYear(year): #if month is feb and leap year its 29 days, otherwise DayInMonth (DiM) is the Jth value in the vector
                days_in_month = 29
            else:
                days_in_month = DiM[j]
            
            for k in range(days_in_month): #loopable through the indexed vector that changes based on state of "J" in previous for loop
                
                file_name = f'{file_base}-{year:4d}-{(j+1):02d}-{(k+1):02d}{file_ext}' #print f-string with zero padding
                files.append(file_name) #add to the end of "files" data structure
                
        year = year + 1 #proceed to loop through next year
    
    return files #return completed list of files


In [2]:
StartYr = 1998 
EndYr = 2000
file_list = ReturnFileList('precip','.nc',StartYr,EndYr) #define the prefix and suffix parameters, set the boundaries for years for the function
#Note that the function is being stored in a variable so that we can retrieve information about the completed function, like how long the list is
print(str(len(file_list))+' files') #store length of list as a string
print(*file_list, sep='\n') #print each value in the list in a new row



1096 files
precip-1998-01-01.nc
precip-1998-01-02.nc
precip-1998-01-03.nc
precip-1998-01-04.nc
precip-1998-01-05.nc
precip-1998-01-06.nc
precip-1998-01-07.nc
precip-1998-01-08.nc
precip-1998-01-09.nc
precip-1998-01-10.nc
precip-1998-01-11.nc
precip-1998-01-12.nc
precip-1998-01-13.nc
precip-1998-01-14.nc
precip-1998-01-15.nc
precip-1998-01-16.nc
precip-1998-01-17.nc
precip-1998-01-18.nc
precip-1998-01-19.nc
precip-1998-01-20.nc
precip-1998-01-21.nc
precip-1998-01-22.nc
precip-1998-01-23.nc
precip-1998-01-24.nc
precip-1998-01-25.nc
precip-1998-01-26.nc
precip-1998-01-27.nc
precip-1998-01-28.nc
precip-1998-01-29.nc
precip-1998-01-30.nc
precip-1998-01-31.nc
precip-1998-02-01.nc
precip-1998-02-02.nc
precip-1998-02-03.nc
precip-1998-02-04.nc
precip-1998-02-05.nc
precip-1998-02-06.nc
precip-1998-02-07.nc
precip-1998-02-08.nc
precip-1998-02-09.nc
precip-1998-02-10.nc
precip-1998-02-11.nc
precip-1998-02-12.nc
precip-1998-02-13.nc
precip-1998-02-14.nc
precip-1998-02-15.nc
precip-1998-02-16.nc
pr

Your Markdown answers go here...

1) The code uses a function (IsLeapYear())to evaluate if a year is a leap year or not. Then, it defines a function (ReturnFileList()) to assemble data files that concatenate file_base, file_ext, and year based on a series of for loops. These loops loop through years, months, and days, with a special consideration for defining February as a 29-day month when the leap year function returns TRUE. The code then applies all of these loops in into the file_name variable, which is an fstring of the prefix, looped variables, and suffix. It adds the file names to a list, and the following code chunk evaluates and prints how long that list is, and then prints all of the file names.

3) The code works, but is marginally incorrect. I think the IsLeapYear determination is a bit incorrect, and should be evaluating if years are divisible by 400 instead of 500. However, this doesn't really matter because we dont have data from years where that is happening (like 2400), and the calculation is robust enough to handle what we forseeably need it for.
It effectively handled adding an extra day to February in 2000, and we know that it will correctly identify 2100 as not a leap year. 

I would change how DiM was referenced, and I prefer the dictionary approach so that it is clearer how each element in the vector is being used, but that is personal preference for clarity. 

If the print statements at the end are just for confirmation, instead of what will be used for extracting the data, it would be much simpler to print the 1st and 29th days of each month and year alongside the # of records, which would show when leap years were handled correctly and that it iterated through all possible days (or even just the 29th day of every month, checking only 34 file names). I am willing to wager that if we know that the code produced all of the right filenames, there is a pipeline to extract all of the files in the files[] list we want without having to print all 1096 of them!

Otherwise, commenting is the main addition for clarity. Using [i] (which could just be [yr] and [j] (which could just be [mo] as dummy counting variables is less descriptive than I would like, but the functions are working where they need to!